cloudmarker/test/test_gcpcloud.py

```diff
@@ -0,0 +1,143 @@
+"""Tests for gcpcloud plugin."""
+
+
+import copy
+import unittest
+from unittest import mock
+
+from cloudmarker.clouds import gcpcloud
+
+mock_firewall_data_no_nextPageToken = {
+    'kind': 'compute#firewallList',
+    'items': [
+        {
+            'priority': 42,
+            'direction': 'ingress',
+            'sourceRanges': [
+                '0.0.0.0',
+            ]
+        }
+    ]
+}
+
+mock_instance_data_no_nextPageToken = {
+    'kind': 'compute#instanceList',
+    'items': [
+        {
+            'status': 'str',
+            'guestAccelerators': [
+                {
+                    'acceleratorCount': 42,
+                    'acceleratorType': 'str',
+                },
+            ]
+        }
+    ],
+    'id': 'str',
+    'selfLink': 'str'
+}
+
+
+class GCPCloudTest(unittest.TestCase):
+    """Tests for gcpcloud plugin."""
+
+    @mock.patch('cloudmarker.clouds.gcpcloud.discovery')
+    @mock.patch('cloudmarker.clouds.gcpcloud.service_account')
+    @mock.patch('builtins.open', mock.mock_open(
+        read_data='{"project_id": "foo"}'))
+    def test_firewall_without_next_page_token(
+            self,
+            mock_service_account,
+            mock_discovery):
+        mock_compute_resource = mock.MagicMock()
+        mock_firewall = mock.MagicMock()
+        mock_list = mock.MagicMock()
+
+        mock_firewall.list.return_value = mock_list
+        mock_list.execute.return_value = mock_firewall_data_no_nextPageToken
+        mock_compute_resource.firewalls.return_value = mock_firewall
+        mock_discovery.build.return_value = mock_compute_resource
+
+        cloud = gcpcloud.GCPCloud('', '')
+        # Consume the data from the generator
+        list(cloud.read())
+
+        mock_list.execute.assert_called_once_with()
+
+    @mock.patch('cloudmarker.clouds.gcpcloud.discovery')
+    @mock.patch('cloudmarker.clouds.gcpcloud.service_account')
+    @mock.patch('builtins.open', mock.mock_open(
+        read_data='{"project_id": "foo"}'))
+    def test_firewall_with_next_page_token(
+            self,
+            mock_service_account,
+            mock_discovery):
+        mock_compute_resource = mock.MagicMock()
+        mock_firewall = mock.MagicMock()
+        mock_list = mock.MagicMock()
+
+        mock_firewall_data = copy.deepcopy(mock_firewall_data_no_nextPageToken)
+        mock_firewall_data.update({'nextPageToken': 'bar'})
+
+        mock_firewall.list.return_value = mock_list
+        mock_list.execute.side_effect = [mock_firewall_data,
+                                         mock_firewall_data_no_nextPageToken]
+        mock_compute_resource.firewalls.return_value = mock_firewall
+        mock_discovery.build.return_value = mock_compute_resource
+
+        # Consume the data from the generator
+        cloud = gcpcloud.GCPCloud('', '')
+        list(cloud.read())
+        self.assertEqual(mock_list.execute.mock_calls,
+                         [mock.call(), mock.call()])
```

Comment:
This code still has a lot of intermediate mocks. Would you like to remove the intermediate mocks to simplify the code? An example I had provided earlier was this:

```
        mock_firewall_data = copy.deepcopy(mock_firewall_data_no_nextPageToken)
        mock_firewall_data.update({'nextPageToken': 'bar'})

        mock_execute = mock_discovery.build().firewalls().list().execute
        mock_execute.side_effect = [mock_firewall_data,
                                    mock_firewall_data_no_nextPageToken]

        # Consume the data from the generator
        list(gcpcloud.GCPCloud('', '').read())

        self.assertEqual(mock_execute.mock_calls, [mock.call(), mock.call()])
```

So with a code like this, we do not need to assign all the intermediate mocks for `mock_firewall.list.return_value`, `mock_compute_resource.firewalls.return_value`, and `mock_discovery.build.return_value` ourselves because the `unittest.mock` does that for us for free when we call `mock_discovery.build().firewalls().list()` above.